tasks.go

```diff
@@ -1,125 +1,121 @@
-	}
-
-	if len(projects) == 0 {
-		return nil, errors.New("no projects visible to the currently logged in user")
-	}
+// NewError returns a Task that always return the given error.
+func NewError(err error) Task {
+	return func() error { return err }
+}
 
-	// Add tasks to fetch resource definitions.
-	definitionsTasks, err := GetResourceDefinitionsTasks(projects, resources, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, definitionsTasks...)
+var (
+	resources = []string{"deploymentconfigs", "pods", "services", "events"}
+	// We should only care about logs for pods, because they cover all other
+	// possible types.
+	resourcesWithLogs = []string{"pods"}
+)
 
-	// Add tasks to fetch logs.
-	logsTasks, err := GetFetchLogsTasks(projects, resourcesWithLogs, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, logsTasks...)
+// GetAllTasks returns a list of all tasks known to the dump tool.
+// FIXME: GetAllTasks should not need to know about basepath.
+func GetAllTasks(basepath string) <-chan Task {
+	tasks := make(chan Task)
+	go func() {
+		defer close(tasks)
 
-	nagiosDataTasks, err := GetNagiosTasks(projects, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, nagiosDataTasks...)
+		projects, err := GetProjects()
+		if err != nil {
+			tasks <- NewError(err)
+			return
+		}
+		if len(projects) == 0 {
+			tasks <- NewError(errors.New("no projects visible to the currently logged in user"))
+			return
+		}
 
-	// Add check tasks
-	for _, p := range projects {
-		outFor := outToFile(basepath, "json", "analysis")
-		errOutFor := outToFile(basepath, "stderr", "analysis")
-		task := CheckTasks(p, outFor, errOutFor)
-		tasks = append(tasks, task)
-	}
+		var wg sync.WaitGroup
+
+		// Add tasks to fetch resource definitions.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			GetResourceDefinitionsTasks(projects, resources, basepath, tasks)
+		}()
+
+		// Add tasks to fetch logs.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			GetFetchLogsTasks(projects, resourcesWithLogs, basepath, tasks)
+		}()
+
+		// Add tasks to fetch Nagios data.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			GetNagiosTasks(projects, basepath, tasks)
+		}()
+
+		// Add check tasks.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for _, p := range projects {
+				outFor := outToFile(basepath, "json", "analysis")
+				errOutFor := outToFile(basepath, "stderr", "analysis")
+				tasks <- CheckTasks(p, outFor, errOutFor)
+			}
+		}()
 
-	if len(retErrors) > 0 {
-		return tasks, retErrors
-	}
-	return tasks, nil
+		wg.Wait()
+	}()
+	return tasks
 }
 
 // GetNagiosTasks will return an array of Tasks each of which will dump the nagios data for one project.
-func GetNagiosTasks(projects []string, basepath string) ([]Task, error) {
-	var tasks []Task
-	var errors errorList
+func GetNagiosTasks(projects []string, basepath string, tasks chan<- Task) {
```

Comment:
The changes to the various functions here should be similar.

In essence, instead of returning `[]Task` and `error`, we take a `chan<- Task` as input.
We send tasks through the channel and they can be immediately run, as opposed to accumulating tasks in a list.

Given that running a task returns an error, and we often proceed even in the presence of errors, instead of taking a channel of errors as input, we simply send errors down the task channel as tasks that "always fail".

Having a single channel in this case facilitates synchronization.